### ISSUE TITLE
Only one packet in the graph

### ISSUE THREAD

**Issue by rriggio** [kohler/click#350]
Hi,

I am using click over wifi and it seems that only one packet is in the graph at a given time. Consider the following configuration:

```
 FromHost(empower0, 192.168.100.100/8)
  -> Print()
  -> queue :: Queue(50)
  -> WifiEncap(0, 00:15:8A:02:DC:6A)
  -> WifiSeq()
  -> RadiotapEncap()
  -> ToDevice(moni0, BURST 1000)
```

From the command line I am trying to transmit some traffic using iperf (after adding an arp entry). I see that the traffic is injected into the wireless interface.

However if I check the length of the Queue element I only see a single packet. Even if I try to inject at a very high rate packets are always dropped at the tun interface and not by the Queue element. After debugging a bit the ToDevice code I noticed that the pcap inject call is blocking the entire router.

Would multithreading help in this situation (userlevel)?

**Comment by pallas** [comment]
I'd have to check to see if userland does the same thing, but in kernel click using handlers blocks the entire router. Since your burst size is 1000 and the Queue length is 50, I would expect the Queue to be mostly empty after any ToDevice run, up to the burst size of FromHost. YMMV.

Multithreading won't help here, since the packets are being dropped outside of click. You'll just send packets out of order. My wild guess is that the tun interface is not signaling backpressure to ToDevice correctly, which is why your Queue never appears to fill.

**Comment by tbarbette** [comment]
Moreover most of those elements are not thread-safe, userlevel or not... More than reorder, I would expect a plain and simple crash.

Drop wireless, use cables.

**Comment by ahenning** [comment]
Does ToDevice(METHOD LINUX) produce the same result as the default METHOD PCAP?

**Comment by tbarbette** [comment]
PCAP should be faster as libpcap evolved with recent techniques, such as PACKET_MMAP for example while LINUX is still the plain old basic socket. But functionally, it's the same.

Not sure if the OP question was currently with kernel? If so, one big difference when coming from kernel is that the FromDevice do not stop packets by default in userlevel, they continue to the normal stack. You only get a copy in click. SNIFFER 1 will insert a rule to stop "the original" from continuing in the stack.

**Comment by rriggio** [comment]
Forgot to mention that I have tried both with burst 1 (the default) and burst 1000. The result is the same: packets are dropped on the tun interface and not on the queue element.

Tried also with METHOD LINUX and the behaviour is the same.

I am using click in userspace to inject traffic on a monitor interface (with radiotap headers). The configuration is more complex however the basic problem is present also in the simple example above.

I have modified the script to use SNIFFER 1, but I get still the same result:

```
ControlSocket(TCP, 7777)

FromDevice(moni0, SNIFFER 1)
-> Discard();

 FromHost(empower0, 192.168.100.100/8)
  -> Print()
  -> queue :: Queue(50)
  -> WifiEncap(0, 00:15:8A:02:DC:6A)
  -> WifiSeq()
  -> RadiotapEncap()
  -> ToDevice(moni0)
```

Why is the tun device dropping the packets instead of passing them downstream?

**Comment by rriggio** [comment]
Turns out this is not related with the wireless. In fact if I run the following script:

```
ControlSocket(TCP, 7777)

FromHost(empower0, 192.168.100.100/8)
 -> queue :: Queue(50)
 -> ToDevice(eth0)
```

and then I run

```
ip neighbor add 192.168.100.20 lladdr 11:22:33:44:55:66 nud permanent dev empower0
iperf -c 192.168.100.20 -u -b 100M
```

I get only max one packet in the queue (checked the handler queue.highwater_length at the end of the transmission)

**Comment by tbarbette** [comment]
And if you put a burst of 32 in FromHost with that last configuration, it does not change a thing?

**Comment by rriggio** [comment]
BURST is supported only in kernelspace. I am using the userspace driver.

**Comment by bcronje** [comment]
Roberto I'm assuming if you use normal FromDevice with ethernet interface you do not experience this issue? I.e. issue is related to FromHost?

Have you tried KernelTun http://read.cs.ucla.edu/click/elements/kerneltun or KernelTap http://read.cs.ucla.edu/click/elements/kerneltap as a replacement to FromHost? That is what I use on a regular basis and havent had any issues with them.

**Comment by tbarbette** [comment]
Ok then try using KernelTun element, it supports burst. It should work the same way than FromHost which also use TUN interfaces in backstage.

This explains why you see that behavior. This element is only reading one packet per "scheduling loop", while ToDevice can probably send more. So the queue is always filled with only one packet always read without any problem by ToDevice.

Once it works, If you need to make the Kernel->Userlevel transfer faster, it may be worth looking at DPDK using the KNI PMD, which use a module in the kernel to span a virtual device that can avoid the kernel to userspace copy (but still needs one internally). It is supposed to be faster than TUN/TAP, though I cannot find any measurements to prove this.

**Comment by rriggio** [comment]
Excellent. Now the traffic is indeed dropped at the queue element and the tun interface reports 0 dropped packets. Thanks for your support.

**Comment by pallas** [comment]
Oh, my mistake. I thought you meant they were being dropped downstream of click.

**Comment by kohler** [comment]
Reopening: perhaps there's a bug in FromHost

**Comment by tbarbette** [comment]
FromHost overlaps a lot with KernelTun. We may suppress it telling somewhere that KernelTun is a suitable userlevel equivalent for FromHost, or just build a wrapper that extends KernelTun? And take advantage of the better code of KernelTun, including burst but not only?

**Comment by bcronje** [comment]
There actually is already a note in FromHost docs: "Note: The KernelTun and KernelTap elements are preferable for most purposes."

Wrapper sounds like a good idea to me, that is what FromSocket/ToSocket does, just extends Socket element.

**Comment by ahenning** [comment]
One benefit of FromHost is that the IP Address on the virtual interface is not mandatory. Other than that I agree with the above.

**Comment by rriggio** [comment]
Agree, the only missing "feature" in KernelTun/Tap is to make the interface address not mandatory. Did not check the code so I am not sure how easy it is to remove this requirement.
